projects/m2/025-checking-for-a-winning-card/js/index.js

```diff
@@ -0,0 +1,161 @@
+ * @param {Array} markedBingoCard the bingo Card to check
+ * @return {boolean} true if a line of 5 zeros is found, false otherwise
+ */
+function isWinning(markedBingoCard){
+    const gameName = 'BINGO';
+    let lineType = '';
+    let won; // This uninitialized control boolean is used to determine if the winning line has been found, it will be set to true at the start of each cycle and to false whenever a non-zero number is found.
+
+    for (let i = 0; i < gameName.length; i++) {
+        // The first check is done for horizontal lines
+        won = true;
+        for (let j = 0; j < gameName.length; j++) {
+            if(markedBingoCard[i][gameName[j]] !== 0){
+                won = false;
+                break;
+            }
+        }
+        // If no winning line is found among the horizontal ones then the vertical ones are checked
+        if(won === false){
+            won = true;
+            for (let j = 0; j < gameName.length; j++) {
+                if(markedBingoCard[j][gameName[i]] !== 0){
+                    won = false;
+                    break;
+                }
+            }
+            // If no winning line is found among the vertical ones then the two diagonals are checked
+            // The first diagonal starts at the top left and ends at the bottom right
+            if(won === false){
+                // To avoid further cycles, the diagonal check is performed only on the main first cycle
+                if(i === 0){
+                    won = true;
+                    for (let j = 0; j < gameName.length; j++) {
+                        if(markedBingoCard[i+j][gameName[j]] !== 0){
+                            won = false;
+                            break;
+                        }
+                    }
+                    // The second diagonal starts at the top right and ends at the bottom left
+                    if(won === false){
+                        won = true;
+                        for (let j = 0; j < gameName.length; j++) {
+                            if(markedBingoCard[i+j][gameName[gameName.length -j -1]] !== 0){
+                                won = false;
+                                break;
+                            }
+                        }
+
+                        if(won === true){
+                            lineType = 'diagonal (from right to left)';
+                            break;
+                        }
+                    } else {
+                        lineType = 'diagonal (from left to right)';
+                        break;
+                    }
+
+                }
+                
+            } else {
+                lineType = 'vertical';
+                break;
+            }
+        } else {
+            lineType = 'horizontal';
+            break;
+        }
+    }
+
+    if(won === true){
+        console.log('The winning line is ' + lineType + '.');
+    }
+    return won;

```

Comment:
The code **needs to be optimized**. 
The first step could be to implement **functions** that specifically deal with checking a result (**horizontal**, **vertical**, or **diagonal**).
It's **not good practice** to do _everything within a single function_.

Reply:
Ok, it's been optimized.